Sprint 13

LobovVit opened the pull request:
Sprint 13

LobovVit committed:
refactor folders

LobovVit committed:
Alert refactor

LobovVit committed:
sp_13  WebView test

LobovVit committed:
ImagesListViewPresenter + ProfileViewController + tests

LobovVit committed:
little changes

LobovVit committed:
sprint_13 UI test

LobovVit committed:
-

NaR0RoW commented on ImageFeed/Alert/AlertModel.swift:
```diff
@@ -0,0 +1,18 @@
+//
+//  Alert.swift
```
Не соответствует названию файла

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeed/Alert/AlertModel.swift:
```diff
@@ -0,0 +1,18 @@
+    let message: String
+    let buttonText: String
+    let completion: (() -> Void)?
+    var secondButtonText: String? = nil
```
Можно не присваивать `nil`

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeed/Auth/AuthConfiguration.swift:
```diff
@@ -0,0 +1,29 @@
+//  ImageFeed
+//
+//  Created by Vitaly Lobov on 16.12.2024.
+//
+import Foundation
```
Не хватает переноса строки

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeed/Auth/AuthHelper.swift:
```diff
@@ -0,0 +1,53 @@
+//  ImageFeed
+//
+//  Created by Vitaly Lobov on 16.12.2024.
+//
+import Foundation
```
Лучше отделять импорты от заголовка файла. Проверь пожалуйста это во всем коде

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeed/Auth/AuthHelper.swift:
```diff
@@ -0,0 +1,53 @@
+import Foundation
+
+protocol AuthHelperProtocol {
+    func authRequest() -> URLRequest?
+    func code(url: URL) -> String?
```
В общем случае функция предполагает выполнение определенного действия, поэтому название этих методов можно немного улучшить, например:
`func createAuthRequest() -> URLRequest?`
Хотя в данном случае я бы воспользовался вычисляемой переменной:
```
var authRequest: URLRequest? {
   guard let url = authURL() else { return nil}
   return URLRequest(url: url)
}
```

Для второго метода, что-то вроде:
`func getCode(for url: URL) -> String?`

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeed/Auth/AuthHelper.swift:
```diff
@@ -0,0 +1,53 @@
+
+final class AuthHelper: AuthHelperProtocol {
+    
+    let configuration: AuthConfiguration
```
Поле можно сделать приватным

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeed/Auth/AuthHelper.swift:
```diff
@@ -0,0 +1,53 @@
+        return URLRequest(url: url)
+    }
+    
+    func authURL() -> URL? {
```
Лучше сделать вычисляемой переменной или изменить название метода на `createAuthURL() -> URL?`

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeed/Auth/AuthHelper.swift:
```diff
@@ -0,0 +1,53 @@
+        return URLRequest(url: url)
+    }
+    
+    func authURL() -> URL? {
```
Для достижения [инкапсуляции](https://habr.com/ru/articles/87205/) лучше пометить это поля `private`.
Перепроверь все свойства и методы в классах. Они должны быть максимально инкапсулированы и скрыты внутри класса

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeed/Auth/AuthHelper.swift:
```diff
@@ -0,0 +1,53 @@
+    func code(url: URL) -> String? {
+        if
+            let urlComponents = URLComponents(string: url.absoluteString),
+            urlComponents.path ==  Constants.authorizeNative ,
```
Лишний пробел перед запятой в конце

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -8,16 +8,19 @@ import UIKit
+    func displayProfileData(name: String?, loginName: String?, bio: String?)
+}
+
+class ProfileViewController: UIViewController, ProfileViewControllerProtocol {
```
Давай проставлять ключевое слово `final` классам для ускорения работы компилятора.

NaR0RoW commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -127,15 +123,16 @@ class ProfileViewController: UIViewController {
     }
     
-    private func updateAvatar() {
+    internal func updateAvatar() {
```
Внутренний уровень стоит по умолчанию для свойств и методов и предоставляет доступ внутри модуля. Явно указывать internal не нужно

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeedTests/ImagesListTests/ImagesListViewControllerSpy.swift:
```diff
@@ -0,0 +1,17 @@
+final class ImagesListViewControllerSpy: ImagesListViewControllerProtocol {
+    
+    var presenter: ImagesListViewPresenterProtocol?
+    var tableViewUpdatesCalled: Bool =  false
```
Лишний пробел после знака равно

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeedUITests/ImageFeedUITests.swift:
```diff
@@ -7,37 +7,82 @@
 
-final class ImageFeedUITests: XCTestCase {
-
+class Image_FeedUITests: XCTestCase {
```
Не хватает `final`

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeedUITests/ImageFeedUITests.swift:
```diff
@@ -7,37 +7,82 @@
+    
+    private let app = XCUIApplication() // переменная приложения
+    
+    private enum const {
```
Название `enum` должно начинаться с большой буквы

LobovVit replied:
Подправил

NaR0RoW commented on ImageFeedUITests/ImageFeedUITests.swift:
Комментарии можно убрать

LobovVit committed:
changes by comments (sprint_13)

LobovVit committed:
clear comments

NaR0RoW reviewed (approved):
По коду все хорошо, отличная работа! Теперь нужно поправить тесты.

Еще такой момент: если поправил мой комментарий в пул реквесте, то достаточно нажать кнопку `Resolve conversation`. А вот если остались вопросы или мой комментарий не получается исправить, то кнопку `Resolve conversation` нажимать не нужно и стоит написать свой комментарий в треде

LobovVit committed:
fix: UI Tests

LobovVit committed:
fix: UI test (testFeed) v2

NaR0RoW commented on ImageFeed/Images/ImagesListCell.swift:
```diff
@@ -50,7 +50,7 @@ final class ImagesListCell: UITableViewCell {
             dateLabel.text = ""
         }
-        
+        likeButton.accessibilityIdentifier = "LikeButton"
```
У нас ведь уже есть `accessibilityIdentifier` у `LikeButton`, поставленный в `Main Storyboard`

NaR0RoW commented on ImageFeedUITests/ImageFeedUITests.swift:
```diff
@@ -45,32 +45,29 @@ final class Image_FeedUITests: XCTestCase {
-        let cellToLike = tablesQuery.children(matching: .cell).element(boundBy: 0)
-        XCTAssertTrue(cellToLike.waitForExistence(timeout: 3))
         
+        let cell = tablesQuery.children(matching: .cell).element(boundBy: 0)
+        cell.swipeUp()
```
Тут стоит ждать, пока `cell` появится на экране. У нас ведь `Splash Screen` до этого показывается

NaR0RoW commented on ImageFeedUITests/ImageFeedUITests.swift:
```diff
@@ -45,32 +45,29 @@ final class Image_FeedUITests: XCTestCase {
+        cell.swipeUp()
+        sleep(4)
+        
+        let cellToLike = tablesQuery.children(matching: .cell).element(boundBy: 1)
```
Мы делаем свайп и пытаемся нажать на лайк, который не видно на экране. У меня тест все равно падает на этом моменте

NaR0RoW commented:
Видео на две части обрезал, т.к. гитхаб не позволяет больше 10 МБ за раз загрузить

https://github.com/user-attachments/assets/5b259659-9f6c-4439-95a6-ff753c2d78e2

https://github.com/user-attachments/assets/b3cce74a-34d9-4fc3-afea-b4e2b5ac6d45

LobovVit committed:
little change

LobovVit commented:
у меня такое поведение при медленной сети 
если все ОК с новым ключом и без rate limit от Unsplash - https://disk.yandex.ru/i/QHhVUqr0quZD2g

пробовал проверять на isHittable по сути все элементы и кнопку и ячейку - и крутил в цикле от первого до 5-го элемента  -  ошибка у меня вылетает периодически (если смотреть на тест - то вылетает когда ProgressHUD крутится долго - дольше тайм-аутов ожидания)

NaR0RoW commented:
С сетью у меня проблем нет, разве что с сервисом могут быть проблемы

<img width="1069" alt="Screenshot 2024-12-19 at 19 38 36" src="https://github.com/user-attachments/assets/fd3e0dda-fb98-4f7d-8a20-03d30c55d31d" />

Вот тут например видно, `ProgressHUD` прокрутился достаточно быстро, но тест все равно упал

https://github.com/user-attachments/assets/a8f1bde4-7cf1-4ff8-915b-be9455c23b70

NaR0RoW commented:
И такое поведение всегда повторяется. На 11 айфоне тоже. Xcode 16.2 iOS 18.2

LobovVit committed:
fix: UI test v3

LobovVit committed:
Крайние правки - дальше не понимаю отчего может еще свалиться

LobovVit committed:
fix: (UI test)
убрал sleep между тапами лайков - ошибка стала проявляться гораздо реже (у меня из 20 запусков только один раз)

NaR0RoW commented:
https://github.com/user-attachments/assets/399532c3-cda5-465c-953c-6734a07b9ba7

https://github.com/user-attachments/assets/eec1eaf6-08a2-44ff-94de-b73339619178

NaR0RoW commented:
https://github.com/user-attachments/assets/6e061178-7b6c-420f-8176-2ad8abd34833


https://github.com/user-attachments/assets/0a0f2395-8605-4b10-9f75-7bb9df1dc89c


https://github.com/user-attachments/assets/245af004-3c7d-47e1-b197-25c0a3c00582

NaR0RoW commented:
Привет!
К сожалению, все еще не могу принять твою работу, так как тест у меня всегда падает. Обратись пожалуйста к наставнику за помощью.

LobovVit committed:
fix (UI test)

NaR0RoW reviewed (approved):
Теперь все работает как нужно, отличная работа 🔥
